Dropping phedex args

jhonatanamado opened the pull request:
Fixes #11145 and partially #9639 (Only T0 system)

#### Status
Tested with a replay

#### Description
Changes in several wfs arguments, some DAOs and RucioInjector to completely drop most of PhEDEx arguments in dataset subscriptions and use some of those arguments as an additional information for Rucio subscriptions 
To wrap-up the following modifications are included in this PR:

- Delete from the databases the following entries. `move`, `phedex_group `
- Delete from the subscriptionInfo and wmspec `CustodialSubType `, `NonCustodialSubType `, `CustodialGroup ` ,`NonCustodialGroup ` and `AutoApproveSubscriptionSites`
- Map the priority from the subscriptionInfo to priority for Rucio Container rules


#### Is it backward compatible (if not, which system it affects?)
No. Database schema change with this dev

#### Related PRs
[T0#4679](https://github.com/dmwm/T0/pull/4679)
#### External dependencies / deployment changes
<Does it require deployment changes? Does it rely on third-party libraries?>
No

amaltaro commented:
test this please

jhonatanamado commented:
Hi @amaltaro. Alan, It remains only one thing to this PR and is related to map the priority from the string value to the integer value. With that in mind where this mapping should take place? Should be something like an easy dict within the `insertContainerRules` function described like the following:
```
priority_map={'low'=2,
              'normal'=3,
              'high'=4,
              'reserved='5'}
```

And map those values with something like `ruleKwargs['priority']=priority_map.get(subInfo['priority'])` .

@vkuznet , @todor-ivanov . Guys feel free to start review this PR.

amaltaro reviewed (commented):
This is looking good, Jhonatan.

Regarding the priority, I think I would add that map of string priority to scalar integer in the RucioInjectorPoller (through a method/function). Another option would be to implement a `convertRulePriority` - or something like that - in the Services/Rucio module, and simply call it in RucioInjectorPoller as well. In short, I would leave things in the database as they are and only map the priority right before making a rule.

I should be able to change the Oracle-based Jenkins test tomorrow, such that we can have a full run of these changes with that backend flavor as well.

amaltaro commented:
For the record, I executed the full suite of tests in Jenkins with Oracle and apparently there are no new unit test failures.

jhonatanamado commented on src/python/WMComponent/RucioInjector/RucioInjectorPoller.py:
```diff
@@ -489,6 +489,7 @@ def insertContainerRules(self):
             rseName = subInfo['site'].replace("_MSS", "_Tape")
             container = subInfo['path']
             lifetime = subInfo['dataset_lifetime']
+            logging.info("subInfo %s",subInfo)
```
Yes was a leftover to check the finial Info.

jhonatanamado commented:
I still need to test the last changes related to mapping the PhEDEx priority to rucio priority. I will prepare a replay and let you know the results.

jhonatanamado commented:
Hi @amaltaro Alan, while testing the last [changes](https://github.com/dmwm/WMCore/pull/11181/commits/bcacb79ffbc4e859c856a8611ef6ab8bee7965fd) , Im getting an error from the agent with the following message:
```
    * NoJobReport (Exit Code: 99303)
        * Could not find jobReport
        * 
        * ========== condor.75364.68.err ==========
        * Adding end of condor.75364.68.err to error message:
        * 
        * Traceback (most recent call last):
        *   File "Startup.py", line 17, in <module>
        *     import WMCore.WMRuntime.Bootstrap as Bootstrap
        *   File "<frozen zipimport>", line 259, in load_module
        *   File "/srv/job/WMCore.zip/WMCore/WMRuntime/Bootstrap.py", line 23, in <module>
        *   File "<frozen zipimport>", line 259, in load_module
        *   File "/srv/job/WMCore.zip/WMCore/WMRuntime/TaskSpace.py", line 17, in <module>
        *   File "<frozen zipimport>", line 259, in load_module
        *   File "/srv/job/WMCore.zip/WMCore/WMSpec/WMWorkload.py", line 20, in <module>
        *   File "<frozen zipimport>", line 259, in load_module
        *   File "/srv/job/WMCore.zip/WMCore/WMSpec/WMWorkloadTools.py", line 21, in <module>
        *   File "<frozen zipimport>", line 259, in load_module
        *   File "/srv/job/WMCore.zip/WMCore/Services/Rucio/Rucio.py", line 17, in <module>
        * ModuleNotFoundError: No module named 'rucio'
        * + cp 'WMTaskSpace/Report*.pkl' ../
        * cp: cannot stat 'WMTaskSpace/Report*.pkl': No such file or directory
        * + ls -l WMTaskSpace
        * ls: cannot access WMTaskSpace: No such file or directory
        * + ls -l 'WMTaskSpace/*'
        * ls: cannot 
```
The traceback goes until `Rucio.py` with no module name `'rucio'`, but that module is defined. Any insight?

amaltaro commented:
@jhonatanamado Hi Jhonatan, apologies for the delay getting back to this.
The error is strange to me, especially because you simply replaced PhEDEx by Rucio, so the upper package (Services) is still the same. Did you happen to test these changes before you made this priority map?

I could be wrong, but we might need to add a `WMCore.Services.Rucio` dependency to the `wmc-runtime` package, here:
https://github.com/dmwm/WMCore/blob/master/setup_dependencies.py#L33
If we do so, then a new release needs to be built.

BTW, you can proceed with these tests, but before merging these changes, I would rename `RUCIO_SUBSCRIPTION_PRIORITIES` to `RUCIO_RULES_PRIORITY` to avoid confusion.

jhonatanamado commented:
> The error is strange to me, especially because you simply replaced PhEDEx by Rucio, so the upper package (Services) is still the same. Did you happen to test these changes before you made this priority map?

Hi Alan, yes, I tested before that commit and everything was working.

> I could be wrong, but we might need to add a WMCore.Services.Rucio dependency to the wmc-runtime package, here:
https://github.com/dmwm/WMCore/blob/master/setup_dependencies.py#L33
If we do so, then a new release needs to be built.

Does that means that first, a new WMCore release needs to be build including that dependency in order to test the changes in this PR? or I can just add that new dependency and start to test it again.?

> BTW, you can proceed with these tests, but before merging these changes, I would rename RUCIO_SUBSCRIPTION_PRIORITIES to RUCIO_RULES_PRIORITY to avoid confusion.

Sure!

jhonatanamado commented:
> 

Sure if it is the best way to do it Im fine.

jhonatanamado commented:
Hi @amaltaro , Alan is it possible for you to trigger the test from cmsdmwmbot. Im getting the following result if I tested it in my dev area.
For example for `WorkQueue_t.py`
```
Using MockDBSApi
.
----------------------------------------------------------------------
Ran 35 tests in 949.309s

OK
Exception ignored in: <function MockPlugin.__del__ at 0x7f7fd9375280>
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/BossAir/Plugins/MockPlugin.py", line 268, in __del__
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/BossAir/Plugins/MockPlugin.py", line 248, in close
AttributeError: 'NoneType' object has no attribute 'debug'

```

amaltaro commented:
@jhonatanamado that AttributeError exception above is harmless and tests are still succeeding. I tried to fix it in the past, but couldn't find how. Whenever you want to trigger Jenkins tests, please type the same message that I will write in my next comment.

amaltaro commented:
test this please

amaltaro commented:
test this please

amaltaro commented:
test this please

amaltaro commented:
@jhonatanamado Jhonatan, can you please update the initial PR description with further details of what is provided with this PR? Especially focus to what is a) removed; b) added; please. Thanks

jhonatanamado commented:
@amaltaro , If Im not wrong, the following file will not be used anymore if this PR is merged 
[PhEDEx/DataStructs/SubscriptionList.py](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/PhEDEx/DataStructs/SubscriptionList.py,) Should I delete it?

amaltaro commented:
Thanks for checking this, Jhonatan. Yes, if it's no longer used, then we better remove the whole PhEDEx package here:
https://github.com/dmwm/WMCore/tree/master/src/python/WMCore/Services/PhEDEx

make sure you also remove its equivalent unit tests as well (in a separate commit). Thanks

jhonatanamado commented:
test this please

amaltaro commented:
@jhonatanamado I am pretty sure that these unit test errors are related to your changes:
```
    WMCore_t.WMSpec_t.StdSpecs_t.Resubmission_t.ResubmissionTests:testCustomStepChainACDC changed from success to error
    WMCore_t.WMSpec_t.StdSpecs_t.Resubmission_t.ResubmissionTests:testFailReRecoACDC changed from success to error
```

can you please take a look at them?

amaltaro commented on src/python/WMComponent/RucioInjector/Database/MySQL/GetUnsubscribedDatasets.py:
```diff
@@ -49,7 +47,7 @@ def formatToPhEDEx(self, result):
                 entry['request_only'] = 'y'
             del entry['auto_approve']
-            for key in ['custodial', 'move']:
+            for key in ['custodial']:
```
this for loop can be removed and the if statements turned to:
```
                if entry['custodial'] == 1:
                    entry['custodial'] = 'y'
                else:
                    entry['custodial'] = 'n'
```

jhonatanamado replied:
Done with https://github.com/dmwm/WMCore/pull/11181/commits/43a40a017a31b041ecb3a267ba519d2f0d6929f9

amaltaro commented on src/python/WMCore/WMSpec/StdSpecs/StdBase.py:
```diff
@@ -1136,14 +1136,8 @@ def getWorkloadAssignArgs():
-                     "NonCustodialGroup": {"default": "DataOps", "type": str, "assign_optional": True},
                      "SubscriptionPriority": {"default": "Low", "assign_optional": True,
-                                              "validate": lambda x: x.lower() in PhEDEx_VALID_SUBSCRIPTION_PRIORITIES},
+                                              "validate": lambda x: x.lower() in RUCIO_RULES_PRIORITY.keys()},
```
this line could be simply:
```
"validate": lambda x: x.lower() in RUCIO_RULES_PRIORITY},
```

jhonatanamado replied:
Done with https://github.com/dmwm/WMCore/pull/11181/commits/edb3ee5780ddf03634821fa190f0a35db9d51895

amaltaro commented on src/python/WMCore/WMSpec/WMWorkloadTools.py:
```diff
@@ -246,17 +245,8 @@ def validatePhEDExSubscription(arguments):
         if site.endswith('_MSS'):
             raise WMSpecFactoryException("Auto-approval to MSS endpoint is not allowed: %s" % site)
-    if arguments.get("SubscriptionPriority", "Low").lower() not in PhEDEx_VALID_SUBSCRIPTION_PRIORITIES:
+    if arguments.get("SubscriptionPriority", "Low").lower() not in RUCIO_RULES_PRIORITY.keys():
```
Same comment as above, there is no need to explicitly ask for the dictionary keys, so only `not in RUCIO_RULES_PRIORITY` would be enough.

jhonatanamado replied:
Done with https://github.com/dmwm/WMCore/pull/11181/commits/edb3ee5780ddf03634821fa190f0a35db9d51895

amaltaro commented on src/python/WMCore/WMSpec/WMWorkloadTools.py:
```diff
@@ -246,17 +245,8 @@ def validatePhEDExSubscription(arguments):
     for site in arguments.get("AutoApproveSubscriptionSites", []):
```
I think we should take advantage and also drop this parameter `AutoApproveSubscriptionSites`.

jhonatanamado replied:
Done with https://github.com/dmwm/WMCore/pull/11181/commits/f12b5f5c3e9ef2e305b1590c32728cb1a56be26c and https://github.com/dmwm/WMCore/pull/11181/commits/0c5b7562e64ae0de63470d9df5368e2c2d464727

amaltaro reviewed (changes requested):
@jhonatanamado I left a few comments in the code for your consideration. 
Regarding `AutoApproveSites`, I think we should remove it and rely on the requiresApproval logic that we already have implemented in the code. We are making a request schema change now, so we better change everything that we consider necessary and positive for the long run.

Can you please also scan the test json templates that we have here:
https://github.com/dmwm/WMCore/tree/master/test/data/ReqMgr/requests

and fix those relying on the no-longer valid attributes? I think we should have only a couple of them. Thanks

jhonatanamado commented:
test this please

jhonatanamado commented:
> @jhonatanamado I am pretty sure that these unit test errors are related to your changes:

    WMCore_t.WMSpec_t.StdSpecs_t.Resubmission_t.ResubmissionTests:testCustomStepChainACDC changed from success to error
    WMCore_t.WMSpec_t.StdSpecs_t.Resubmission_t.ResubmissionTests:testFailReRecoACDC changed from success to error
can you please take a look at them?

This is kind of weird, I tested on my env and the test finished without issues.
```
.
----------------------------------------------------------------------
Ran 9 tests in 17.630s

OK
```
I will trigger again the Jenkins test to check if the results are the same or something is not quite consistent with the UnitTest results

amaltaro commented:
@jhonatanamado I was wrong. Those errors are related to a race condition of creating/deleting test databases in couchdb. Here is one of those failures:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/13462/testReport/junit/WMCore_t.WMSpec_t.StdSpecs_t.Resubmission_t/ResubmissionTests/testStandardStepChainACDC/

We can disregard those errors, which should be resolved through: https://github.com/dmwm/WMCore/issues/11194

jhonatanamado commented:
test this please

jhonatanamado commented:
Hi @amaltaro , Alan seems that `WMCore_t.WorkQueue_t.WorkQueue_t.` is not working with Jenkins. I run the same test and this is the output. Any ideas?
Meanwhile I will move to update the ReqMgr templates.
```
.
----------------------------------------------------------------------
Ran 35 tests in 998.198s

OK
Exception ignored in: <function MockPlugin.__del__ at 0x7fe5432cf280>
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/BossAir/Plugins/MockPlugin.py", line 268, in __del__
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/BossAir/Plugins/MockPlugin.py", line 248, in close
AttributeError: 'NoneType' object has no attribute 'debug'
```

jhonatanamado committed:
Dropping PhEDEx UnitTest

Dropping phedex StepChain
Dropping phedex WMTask
Dropping phedex WMWorkoad
Dropping phedex WMBSHelper

Delete PhEDEx component unit test


Delete AutoApproval from unit-test


Delete testValidatePhEDExSubscription from WMWorkload_t 


Updating ReqMgr json templates

amaltaro commented:
@jhonatanamado as we quickly discussed this morning, it turns out that we have already been running the oracle jenkins with this patch in.
I fear I am not able to say whether it's failing new unit tests, but it's close to the number of failures we have in our standard/MySQL jenkins unit tests (around 32 failures in MySQL vs 54 in Oracle, but I know some were already failing in Oracle).
I am going to review it a bit later.

amaltaro commented on src/python/WMComponent/RucioInjector/RucioInjectorPoller.py:
```diff
@@ -519,7 +521,6 @@ def insertContainerRules(self):
                     rseName = rseName.replace("cms_type=real", "cms_type=test")
             else:
                 # then it's a T0 container placement
-                ruleKwargs['priority'] = 4
```
I guess you are always defining it in the workflow spec and you no longer need it for T0?

amaltaro commented on src/python/WMCore/WMSpec/WMWorkloadTools.py:
```diff
@@ -243,20 +242,8 @@ def validatePhEDExSubscription(arguments):
-        if site.endswith('_MSS'):
-            raise WMSpecFactoryException("Auto-approval to MSS endpoint is not allowed: %s" % site)
-    if arguments.get("SubscriptionPriority", "Low").lower() not in PhEDEx_VALID_SUBSCRIPTION_PRIORITIES:
+    if arguments.get("SubscriptionPriority", "Low").lower() not in RUCIO_RULES_PRIORITY:
```
this line can be removed, since we already validate it in the StdBase super class.
Actually, can you please review if this method `validatePhEDExSubscription` is actually used/needed somewhere else? I am inclined to say that it can be completely removed.

jhonatanamado replied:
Yes it can be completely removed, I did it with https://github.com/dmwm/WMCore/pull/11181/commits/737ac4de6bef1a3d49e33f70ec16f74955cbdb5f

amaltaro reviewed (changes requested):
@jhonatanamado Jhonatan, these changes are looking great for me. I left a couple of comments along the code though.
In addition to that, could you please squash all these commits into only 2 commits:
1) for src/* changes
2) for test/* changes.

If you are unsure about this step, please let me know because we do not want to lose these changes ;)

jhonatanamado committed:
Dropping PhEDEx args

Dropping PhEDEx args_StdBase
Dropping PhEDEx args_WMTask
Dropping PhEDEx args_WMWorkload


Dropping PhEDEx DAOs

Dropping phedex MySQL_Create
Dropping phedex MySQL_NewSubscription
Dropping phedex Oracle_Create
Dropping phedex Oracle_NewSubscription


Dropping PhEDEx Rucio Component

Dropping phedex UnsubscribedDts
Dropping phedex RucioInjectorPoller
Assign RUCIO_SUBSCRIPTION_PRIORITIES for containerRules 

Add RUCIO_RULES_PRIORITY

Add dict to map PhEDEx priority to Rucio priority

Change RUCIO_RULES_PRIORITY  from Rucio to RucioUtils


Add RUCIO_RULES_PRIORITY to RucioInjectorPoller


Delete PhEDEx component


Drop for loop from custodial


Change to RUCIO_RULES_PRIORITY


Dropping AutoApproveSubscriptionSites


Delete auto_approve-AutoApproveSites NewSubscriptions


Delete autoApproveSites from WMTask


Deletete AutoApproval from WMWorkload


Delete autoapprove from Oracle and create.py


Delete autoapprove from RucioInjector-GetUnsubscr-Dtsets


Delete validatePhEDExSubscription method

jhonatanamado commented:
@amaltaro , Thanks Alan. What about the modifications on the JSON templates, Should those be squashed with src/* changes?

amaltaro commented:
> @amaltaro , Thanks Alan. What about the modifications on the JSON templates, Should those be squashed with src/* changes?

@jhonatanamado no, those changes go together with unit tests (2nd commit)

amaltaro commented:
Tests didn't come back - jenkins seem to be broken - but last unit tests were successful and the latest changes should have been safe enough. Merging it to get a new release out and deployed in testbed.
